Fix build output to solve ESM import issue

robknight opened the pull request:
**See later comments for revised approach.**

According to the ESM standard, ESM imports are meant to use a fully-qualified filename, including the .js extension, when importing a JS module, although some bundlers use import resolution strategies which allow the extension to be omitted. When using Deno or other standards-enforcing tools, an import without an extension will fail.

`tsc` will not add these extensions during transpilation. However, we can add file extensions to the imports in our TypeScript source, and these will be preserved when transpiled to JavaScript. I've added .js extensions to all imports in our `packages` directory, and an `eslint` rule to ensure that we continue to add them to all future imports.

The only downside is that the `.js` extensions look a bit ugly. We might be able to find another solution, but no other option seems obvious right now.

In the course of implementing this, I encountered some out-of-memory errors in GitHub Actions during the `lint` job. I limited concurrency of `lint` tasks to 4, which seems to have fixed it.

artwyman commented:
Seems weird to be including a .js not .ts extension when writing TypeScript code, but if that's what it takes...
A few developer workflow questions on this:
- Does this affect VSCode's ability to understand imports and symbols to enable autocomplete, refactoring, etc?
- How will this interact with auto-imports inserted by VSCode?  Can we configure VSCode to follow this eslint rule, or will it require a manual fixup every time?

ax0 commented:
Not sure about VS Code, but I can confirm that this doesn't affect LSP features in Emacs using `typescript-language-server` via `eglot`.

FlorianDr commented:
> Does this affect VSCode's ability to understand imports and symbols to enable autocomplete, refactoring, etc?

Axel provides configurations and more insights in: 
https://2ality.com/2021/06/typescript-esm-nodejs.html#visual-studio-code

> "javascript.preferences.importModuleSpecifierEnding": "js"
> "typescript.preferences.importModuleSpecifierEnding": "js"

ichub commented:
My understanding is this change is motivated by a requirement for ESM modules to be able to import our packages? 

This might be an ignorant question, but I'm not totally sure I understand why this external requirement has to bleed into our code-base. Couldn't dependents import our packages using `.js`-suffixed package names, without us having to worry about this?

ax0 commented:
> Couldn't dependents import our packages using `.js`-suffixed package names, without us having to worry about this?

The issue with this is that our packages' files will then be parsed and found to include references to non-existent modules due to improper imports and exports. For example, using `@pcd/pod` in Deno yields an error along the lines of
```
error: [ERR_MODULE_NOT_FOUND] Cannot find module '.../node_modules/.deno/@pcd+pod@0.2.0/node_modules/@pcd/pod/dist/esm/src/pod' imported from '.../node_modules/.deno/@pcd+pod@0.2.0/node_modules/@pcd/pod/dist/esm/src/index.js'
```

robknight commented:
The gist of the problem is that we're caught between "standards" and "practices", where the two don't really align with each other.

The ESM standard says that imports need to have fully-qualified paths, so you can't do `import { foo } from "./bar"`, you _must_ do `import { foo } from "./bar.js"`. This is different from the old-school `require()`, which would accept `require("./bar")` and would simply search the file system for `./bar.js`.

_However_, most bundlers work around this by emulating the old-school behaviour even with ESM imports. It violates the standard, but it works. TypeScripts `moduleResolution: "bundler"` tells TypeScript to resolve imports in this way, rather than the stricter standards-based approach.

The problem is that not all bundlers do this, and some newer systems (Deno, and Node's native ESM import) don't use bundlers at all. These are unable to handle import statements that don't include the file extension. This is the cause of the problems both @ax0 and @FlorianDr have encountered.

Therefore, the most-compatible approach is to include the file extensions in the import statements. This can be done in two ways: (1) amend the import statements in the TypeScript so that they look exactly like the imports the emitted JavaScript will contain, which is [what the TypeScript team wants you to do](https://github.com/microsoft/TypeScript/issues/42151), or (2) emit "non-standard" imports and then do some post-processing to add the file extensions. The former is what I've implemented here, but the latter would be easy enough and less disruptive. I'll look at doing that now.

robknight commented:
I've now updated this to use a different approach: a post-build step, which updates the ESM JavaScript build artifacts to use the `.js` extension. This avoids the need to change the TypeScript sources. I've also updated the `test-packaging` scripts, and verified that this change works. As a separate point, we may need to come back to packaging tests and include tests for the scenarios that caused problems here (viz. importing in Deno).

ichub reviewed (approved):
this approach works for me!
